### PR TITLE
use components/codec in components/keys

### PR DESCRIPTION
**Pull request by b41sh**
Signed-off-by: baishen <baishen2009@gmail.com>

###  What have you changed?

This PR switch codec to use components/codec in components/keys.

###  What is the type of the changes?

- Engineering

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

no

###  Does this PR affect `tidb-ansible`?

no

###  Refer to a related PR or issue link (optional)

[5702](https://github.com/tikv/tikv/pull/5702)

### CODE REVIEW

File: components/keys/src/types.rs [hunk 1]
```diff
@@ -34,25 +35,26 @@ impl Key {
...
-        bytes::decode_bytes_in_place(&mut k, false)?;
+        MemComparableByteCodec::try_decode_first_in_place(&mut k)?;
         Ok(k)
     }
 
     /// Gets the raw representation of this key.
     #[inline]
     pub fn to_raw(&self) -> Result<Vec<u8>, codec::Error> {
-        bytes::decode_bytes(&mut self.0.as_slice(), false)
+        let mut k = Vec::with_capacity(self.0.len() / (MEMCMP_GROUP_SIZE + 1) * MEMCMP_GROUP_SIZE);
```
Comment (breeswish) [inline components/keys/src/types.rs]: Looks like you can simply allocate a vector with the same size as the original, to avoid the complex calculation and hide the codec implementation.

File: components/keys/src/types.rs [hunk 2]
```diff
@@ -149,10 +151,10 @@ impl Key {
     pub fn decode_ts_from(key: &[u8]) -> Result<TimeStamp, codec::Error> {
         let len = key.len();
         if len < number::U64_SIZE {
-            return Err(codec::Error::KeyLength);
+            return Err(codec::ErrorInner::KeyLength.into());
         }
-        let mut ts = &key[len - number::U64_SIZE..];
-        Ok(number::decode_u64_desc(&mut ts)?.into())
+        let ts = &key[len - number::U64_SIZE..];
+        Ok(NumberCodec::decode_u64_desc(&ts).into())
```
Comment (breeswish) [inline components/keys/src/types.rs]: Prefer to use the `read_xxx` interface like: `let ts = key.read_u64_desc()?.into();`.

File: components/keys/src/types.rs (hunk 1 above)
Comment (b41sh) [reply components/keys/src/types.rs]: done

File: components/keys/src/types.rs [hunk 3]
```diff
@@ -34,25 +35,26 @@ impl Key {
...
+        encoded.write_comparable_bytes(key).unwrap();
         Key(encoded)
     }
 
     /// Gets and moves the raw representation of this key.
     #[inline]
     pub fn into_raw(self) -> Result<Vec<u8>, codec::Error> {
         let mut k = self.0;
-        bytes::decode_bytes_in_place(&mut k, false)?;
+        MemComparableByteCodec::try_decode_first_in_place(&mut k)?;
```
Comment (breeswish) [inline components/keys/src/types.rs]: You need to shrink the buffer size after decoding.

File: components/keys/src/types.rs [hunk 4]
```diff
@@ -34,25 +35,26 @@ impl Key {
...
         Ok(k)
     }
 
     /// Gets the raw representation of this key.
     #[inline]
     pub fn to_raw(&self) -> Result<Vec<u8>, codec::Error> {
-        bytes::decode_bytes(&mut self.0.as_slice(), false)
+        let mut k = Vec::with_capacity(self.0.len());
+        MemComparableByteCodec::try_decode_first(&self.0.as_slice(), &mut k)?;
```
Comment (breeswish) [inline components/keys/src/types.rs]: This will lead to panic because the **length** of k is `0`. Also you need to shrink the buffer size after decoding.

File: components/codec/src/byte.rs [hunk 5]
```diff
@@ -391,6 +392,51 @@ impl MemComparableByteCodec {
             }
         }
     }
+
+    /// Returns whether `encoded` bytes is encoded from `raw`. Returns `false` if `encoded` is invalid.
+    pub fn is_encoded_from(encoded: &[u8], raw: &[u8]) -> bool {
```
Comment (breeswish) [inline components/codec/src/byte.rs]: Prefer to separate it into different PRs. One to implement `is_encoded_from`, one to switch to `components/codec`.

Comment (b41sh) [reply components/codec/src/byte.rs]: ok

Comment (b41sh) [reply components/codec/src/byte.rs]: this code is copied from [5641](https://github.com/tikv/tikv/pull/5641)
why we don't implement `is_encoded_from` in `components/codec` this PR?
@breeswish

Comment (breeswish) [reply components/codec/src/byte.rs]: @b41sh  implementing it for components/codec is supposed to be done after #5641 .

File: components/keys/src/types.rs (hunk 2 above)
Comment (b41sh) [reply components/keys/src/types.rs]: done

File: components/keys/src/types.rs (hunk 3 above)
Comment (b41sh) [reply components/keys/src/types.rs]: done

File: components/keys/src/types.rs (hunk 4 above)
Comment (b41sh) [reply components/keys/src/types.rs]: done

File: components/codec/src/byte.rs (hunk 5 above)
Comment (b41sh) [reply components/codec/src/byte.rs]: ok

**Comment by breeswish** [comment]
Thanks a lot! A friendly hint: to keep the task easy, I would recommend you to exclude the `is_encoded_from` implementation (since it will take notable time to review) and only change other occurrence. Then your PR can be merged very quickly :)

**Comment by b41sh** [comment]
> Thanks a lot! A friendly hint: to keep the task easy, I would recommend you to exclude the `is_encoded_from` implementation (since it will take notable time to review) and only change other occurrence. Then your PR can be merged very quickly :)

ok, I have removed `is_encoded_from`

File: components/keys/src/types.rs [hunk 6]
```diff
@@ -34,25 +36,33 @@ impl Key {
...
         Key(encoded)
     }
 
     /// Gets and moves the raw representation of this key.
     #[inline]
     pub fn into_raw(self) -> Result<Vec<u8>, codec::Error> {
         let mut k = self.0;
-        bytes::decode_bytes_in_place(&mut k, false)?;
+        MemComparableByteCodec::try_decode_first_in_place(&mut k)?;
+        k.shrink_to_fit();
```
Comment (breeswish) [inline components/keys/src/types.rs]: I don't think shrink_to_fit will work. It only shrinks extra capacity while in this case there is extra data in length.

Comment (b41sh) [reply components/keys/src/types.rs]: `shrink_to(written_bytes)`?

Comment (breeswish) [reply components/keys/src/types.rs]: you may want the truncate method.

**Comment by breeswish** [comment]
Please fix your implementation until unit tests pass. You can run them via `make dev`.

**Comment by b41sh** [comment]
unit tests has passed
@breeswish  PTAL

File: components/codec/src/error.rs [hunk 7]
```diff
@@ -11,6 +11,9 @@ pub enum ErrorInner {
 
     #[fail(display = "Data padding is incorrect")]
     BadPadding,
+
+    #[fail(display = "bad format key(length)")]
+    KeyLength,
```
Comment (breeswish) [inline components/codec/src/error.rs]: Prefer not to introduce this variant, since it is not used anywhere in the codec itself.

File: components/codec/src/error.rs [hunk 8]
```diff
@@ -25,6 +28,16 @@ impl ErrorInner {
     }
 }
 
+impl ErrorInner {
```
Comment (breeswish) [inline components/codec/src/error.rs]: Prefer not to introduce `maybe_clone` since it is not used in codec.

File: components/codec/src/error.rs [hunk 9]
```diff
@@ -50,6 +63,18 @@ impl Fail for Box<ErrorInner> {
 #[fail(display = "{}", _0)]
 pub struct Error(#[fail(cause)] pub Box<ErrorInner>);
 
+impl Error {
+    pub fn maybe_clone(&self) -> Option<Error> {
+        self.0.maybe_clone().map(Error::from)
+    }
+}
+
+impl std::error::Error for Error {
```
Comment (breeswish) [inline components/codec/src/error.rs]: `Error` already derives `Fail`, you can use `Fail` infra and no need to implement `std::Error` for it again.

File: components/keys/src/types.rs [hunk 10]
```diff
@@ -2,10 +2,14 @@ use crate::timestamp::TimeStamp;
 use byteorder::{ByteOrder, NativeEndian};
 use hex::ToHex;
 use std::fmt::{self, Debug, Display, Formatter};
-use tikv_util::codec;
+
+use codec;
```
Comment (breeswish) [inline components/keys/src/types.rs]: This line is not redundant.

File: components/keys/src/types.rs [hunk 11]
```diff
@@ -2,10 +2,14 @@ use crate::timestamp::TimeStamp;
...
+use codec::byte::MemComparableByteCodec;
+use codec::number;
+use codec::prelude::{
+    MemComparableByteDecoder, MemComparableByteEncoder, NumberDecoder, NumberEncoder,
+};
```
Comment (breeswish) [inline components/keys/src/types.rs]: Please refine these imports.

File: components/keys/src/types.rs [hunk 12]
```diff
@@ -35,24 +39,26 @@ impl Key {
...
+            MemComparableByteCodec::try_decode_first_in_place(&mut k)?;
+        k.truncate(written_bytes);
         Ok(k)
     }
 
     /// Gets the raw representation of this key.
     #[inline]
     pub fn to_raw(&self) -> Result<Vec<u8>, codec::Error> {
-        bytes::decode_bytes(&mut self.0.as_slice(), false)
+        self.0.as_slice().read_comparable_bytes()
```
Comment (breeswish) [inline components/keys/src/types.rs]: maybe `self.0.read_comparable_bytes()` is enough?

File: components/keys/src/types.rs [hunk 13]
```diff
@@ -114,7 +120,7 @@ impl Key {
             // functions to convert between `TimestampedKey` and `Key`.
             // `TimestampedKey` is in a higher (MVCC) layer, while `Key` is
             // in the core storage engine layer.
-            Err(codec::Error::KeyLength)
+            Err(codec::ErrorInner::KeyLength.into())
```
Comment (breeswish) [inline components/keys/src/types.rs]: Consider introducing a new `Error` for `Key` so that you can customize what it needs.

File: components/codec/src/error.rs (hunk 7 above)
Comment (b41sh) [reply components/codec/src/error.rs]: done

File: components/keys/src/types.rs (hunk 12 above)
Comment (b41sh) [reply components/keys/src/types.rs]: we can't
because error[E0599]: no method named `read_comparable_bytes` found for type `std::vec::Vec<u8>` in the current scope

**Comment by b41sh** [comment]
@breeswish PTAL
I still have two unresolved issues
remove `maybe_clone`
use `Fail` infra to implement `std::error::Error`?
can you give some examples?
and please review the other parts by the way

**Comment by breeswish** [comment]
Sorry for the late reply. In general you can introduce a new Error type using quick-error or failure for `Key` itself, so that then you can introduce your one error variants for it.

**Comment by andylokandy** [comment]
@breeswish  PTAL

**Comment by andylokandy** [comment]
@b41sh Please fix tests.

**Comment by b41sh** [comment]
@andylokandy @breeswish
tests have fixed.
still some problem with errors, `codec::error::Error` is not implement `std::error::Error`, how to implement it? I need some help. 

```
error[E0277]: the trait bound `codec::error::Error: std::error::Error` is not satisfied
  --> components/txn_types/src/lib.rs:35:19
   |
35 |             cause(err)
   |                   ^^^ the trait `std::error::Error` is not implemented for `codec::error::Error`
   |
   = note: required for the cast to the object type `dyn std::error::Error`
```
